storage/connectors/redis/src/main/java/feast/storage/connectors/redis/writer/RedisCustomIO.java

```diff
@@ -250,35 +211,101 @@ private FailedElement toFailedElement(
         return FeatureRow.newBuilder()
             .setEventTimestamp(featureRow.getEventTimestamp())
             .addAllFields(values)
-            .build()
-            .toByteArray();
+            .build();
       }
 
       @ProcessElement
       public void processElement(ProcessContext context) {
-        List<FeatureRow> featureRows = Lists.newArrayList(context.element().iterator());
-
+        List<FeatureRow> filteredFeatureRows = Collections.synchronizedList(new ArrayList<>());
         Map<String, FeatureSetSpec> latestSpecs =
-            context.sideInput(featureSetSpecsView).entrySet().stream()
-                .map(e -> ImmutablePair.of(e.getKey(), Iterators.getLast(e.getValue().iterator())))
-                .collect(Collectors.toMap(ImmutablePair::getLeft, ImmutablePair::getRight));
+            getLatestSpecs(context.sideInput(featureSetSpecsView));
+
+        Map<RedisKey, FeatureRow> deduplicatedRows =
+            deduplicateRows(context.element(), latestSpecs);
 
         try {
-          executeBatch(featureRows, latestSpecs);
-          featureRows.forEach(row -> context.output(successfulInsertsTag, row));
+          executeBatch(
+              (redisIngestionClient) ->
+                  deduplicatedRows.entrySet().stream()
+                      .map(
+                          entry ->
+                              redisIngestionClient
```

Comment:
As an eventual optimization, would it make sense to cache the value that is returned from Redis? So that if we know that all values that we want to write is older than what is cached, we can then discard it all without even looking at Redis. Might be a premature optimization though.

Reply:
Within batch rows are already deduplicated and on this moment it's only one value per key. So each redis key being fetched only once.
Keep this cache as state between batches might be too expensive.